### PR TITLE
Linux platform SSH process creation for remoting

### PR DESCRIPTION
**Pull request by PaulHigin**
Update:
--------
The technique is to now use sigaction to create an SSH client process that ignores SIGINT signals.  It remains in the PowerShell process group and has access to the terminal, which means it can prompt on the command line as before.  So SSH_ASKPASS dependency is not needed.

-------

This change addresses Issue #2321.  The original idea for this fix was to follow what I did in Windows and create the SSH process in its own process group.  That way a Ctrl+C typed signal in PowerShell (the parent process) wouldn't be handled by the child SSH process (causing it to end), and instead would allow PowerShell to handle it through its protocol (PSRP).

But this ended up disabling the /dev/tty terminal for the SSH process so it could not prompt the user for a password, if that was how SSH was set up.

It turns out that SSH will detect if there is no tty available and fall back to using SSH_ASKPASS program if available.  This will happen if the SSH child process is created in its own session.  So this change does that, it creates the SSH process (used for SSH remoting) in its own session.

The downside to this is that now the SSH_ASKPASS program needs to be installed if we want to support any SSH prompting.  We will need to make sure it gets installed with Enable-SSHRemoting (a separate issue/work item).

Creating a Linux process in its own session does not seem like a common need and so I felt it was not necessary to add this functionality to DotNet CoreFx.  Instead I pulled some of the UNIX process creation code over to PowerShell and added the ability to create in a separate session.  So this change has a fair amount of native and platform interop code.

I will update the license document to indicate that PowerShell is using this DotNet CoreFx code under the MIT license.

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@PaulHigin Please add reviewers for your PR and assign it to an appropriate maintainer.

**Comment by iSazonov** [comment]
I leave a [comment in the original Issue](https://github.com/PowerShell/PowerShell/issues/2321#issuecomment-305687171)

**Comment by PaulHigin** [comment]
@iSazonov Please put comments here rather than providing a link.  The SSH -t option to create a pseudo tty prevents using subsystems which is needed to host PowerShell for SSH remoting.  So it won't work here as far as I can tell.  If there was some way to create and use the child SSH process such that it is not part of the process group (so that it doesn't get Ctrl+C signals) but still can use a tty for user prompt, then that would be great.  But I don't currently see a way to do this.  However, I am definitely open to any ideas as this would be preferable to using SSH_ASKPASS!

**Comment by iSazonov** [comment]
@PaulHigin I don't read in depth but it is interesting https://unix.stackexchange.com/questions/266866/how-to-prevent-ctrlc-to-break-ssh-connection

**Comment by iSazonov** [comment]
Oh, I think about a remote session all the time, but the problem is on the local side, isn't it? If so there is a simple test:
1. run Bash
2. run stty intr undef
3. run PowerShell and test Ctrl-C in remote connection

If this works well we can make the same internally in PowerShell - disable ISIG by means of [TERMIOS(3)](http://man7.org/linux/man-pages/man3/termios.3.html) at startup time.

**Comment by PaulHigin** [comment]
This looks interesting, but I don't understand what the fix would be.  I don't want to disable the ISIG for the PowerShell /dev/tty but instead want to keep the child SSH process from getting it.  I only want PowerShell to handle it.  It is not clear to me that disabling ISIG via termios can be done per process.  But if so that would certainly solve our problem!

BTW I did look at your previous link as a way to turn off ISIG processing on the server side, but I wasn't able to get it to work.  "set -m"  didn't seem to do anything.

**Comment by PaulHigin** [comment]
Hmm, sigaction() looks interesting.  It should be possible to suppress a signal on a specific process.  I'll look into it.

**Comment by PaulHigin** [comment]
Sigaction() works great and obviates the need for SSH_ASKPASS since the SSH process can be created in its parent group/session but still be made to ignore interrupt signals (SIGINT).

I have updated the PR to use sigaction().

**Comment by daxian-dbw** [comment]
Assign this PR to @mirichmo because he would have more context as one of the reviewers.

**Comment by iSazonov** [comment]
@PaulHigin Good catch!  I reread the discussion and it looks that `sigaction`  is exactly what we wanted. 

I believe we should consider [sigprocmask](http://man7.org/linux/man-pages/man2/sigprocmask.2.html) 

Before the code review I still have thoughts how make it easy. Because we search just a workaround for Unix we can: 
1. Use simple wrapper to run "ssh" like [this](https://stackoverflow.com/questions/4515274/externally-disabling-signals-for-a-linux-program) 
2. Ask OpenSSH team to add new option for ignoring SIGINT.
3. If none of the preceding is satisfied we continue the PR as is.

I'd rather not complicate the code and use (1) option.

**Comment by PaulHigin** [comment]
@iSazonov Thanks for you help!
I don't want to use "mask" because the documentation indicates that it masks signals for the entire process group.  Instead I just want the SSH process to ignore the signal.
1. That is what this fix is!  The only difference is that we configure the SSH process start to ignore the signal instead of having the SSH process do it.
2. Or do you mean to move the fix from PowerShell into OpenSSH?  I agree that we should pursue this in the long term but there are a couple of problems:
 - The change may not be accepted or it may take a long time to get accepted as the OpenSSH group must think about wider consequences and compatibility issues.
 - Even if we get the change approved it will still leave many systems with older versions of OpenSSH and without the needed fix.  Fixing this in PowerShell ensures it will just work after installing PowerShell.

The nice thing about open source is that the code can be refactored by anyone if/when a better idea comes along.  But for now I feel this is the best solution.

**Comment by PaulHigin** [comment]
Removing "breaking change" tag since SSH_ASKPASS is no longer needed and behavior remains the same.

**Comment by iSazonov** [comment]
@PaulHigin 
I agree that the old OpenSSH versions will not solve the problem even though new versions get new option to mask SIGINT, but I should have said this option. :-)

I don't see in docs that sigprocmask masks signals for the entire process group - it seems it affects only the [current process](http://man7.org/linux/man-pages/man2/sigprocmask.2.html):
>A child created via fork(2) inherits a copy of its parent's signal
       mask; the signal mask is preserved across execve(2).

Now your code is doing what we need. I suggest to discuss an option that is dear to the Unix world:
- bring the code to external wrapper  `SSHrunner`: to fork - sigprocmask - execvp - ssh
- use our statdard process call to run the wrapper

I believe it's much simpler than use internally CoreFX code. What do you think?

**Comment by SteveL-MSFT** [comment]
I agree with @iSazonov that a wrapper may be simpler than having a copy of the dotnet code.  Should we reuse powershell.exe for this purpose?

**Comment by iSazonov** [comment]
> Should we reuse powershell.exe for this purpose?

I think yes, we can add new "secret" switch for that `powershell -startSSH`.

**Comment by PaulHigin** [comment]
Have two processes to run SSH?  This seems like a waste of processes to me.  For fan-out scenarios it means having twice as many processes on the client for each supported session.  This does not sound like a good design to me.  The code to create a new process is quite small and can be made much smaller if it is meant to be SSH specific.  I just copied the dotNet general process creation code but we don't need all of it.  And I really dislike using powershell.exe to be a process host wrapper since it is pretty heavy weight with the CLR.  I'll work at making the launch code smaller and specific but I feel we should stick with the current process launch design.

**Comment by iSazonov** [comment]
>Have two processes to run SSH? 

No, we run a wrapper - (1) fork and (2) fork to SSH - so we have one parent process - PowerShell, and one child process - SSH. It is not Windows, It is Unix and it is fast and best practice for Unix.

>I really dislike using powershell.exe to be a process host wrapper since it is pretty heavy weight with the CLR

I agree but we can fork in Main(string[] args) before PowerShell initialization code. So it should be fast.

**Comment by PaulHigin** [comment]
Yes, but there is no reason to do it.  Even if the wrapper doesn't "run" it takes up resources, and I assume the scheduler has to account for it.  Also std pipes must be routed.  Why do this when we can easily create a single process?

**Comment by iSazonov** [comment]
>Also std pipes must be routed

Fork inherits its.

I have no more thoughts and PG can make an conclusion.

**Comment by PaulHigin** [comment]
Nevertheless my position remains "why do it when it is not necessary?".  Its funny but I just watched some presentations about performance yesterday.  A big theme from the presenter was that best practices and accepted patterns were many times poor choices performance wise.  One of the top things the presenter looked for when investigating performance problems was unnecessary work such as memory allocations, thread creation, uses of locks, and other system resource usage.  So after that the idea of creating an unneeded process when it is just as easy to create a single process seems undesirable.

**Comment by daxian-dbw** [comment]
Very interesting discussion here and I believe calling `sigactions` before `execve` is the best solution.

I changed to code in `ForkAndExecProcess` a bit on CoreFx to fix issue https://github.com/dotnet/corefx/issues/13447, I think you should use the latest code instead. I will leave a comment there.

File: src/libpsl-native/src/createprocess.cpp
Comment (daxian-dbw) [inline src/libpsl-native/src/createprocess.cpp]: The code in this function was changed a bit to address https://github.com/dotnet/corefx/issues/13447, you can see the latest code at [`pal_process.cpp`](https://github.com/dotnet/corefx/blob/release/2.0.0/src/Native/Unix/System.Native/pal_process.cpp#L142). The change was basically to make all pipe file descriptors close-on-exec. If not doing this here, then it's possible that the ssh process will continue to block on waiting for input even if the redirected standard input has been closed in powershell.

**Comment by iSazonov** [comment]
@PaulHigin @daxian-dbw I want to get clarify about  `sigactions` vs `sigprocmask`. I still belive we can use sigprocmask after fork so OS will not trigger the sigactions empty SIGINT handler in SSH process.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: It seems we can use this everywhere. So maybe put this in "right place" as internal?

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Should we add an usefull message?

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs]: Why we remove ".exe" for Windows? We can catch .bat or .com file.

File: src/libpsl-native/src/createprocess.cpp
Comment (PaulHigin) [inline src/libpsl-native/src/createprocess.cpp]: I don't think this is applicable in this situation.  The pipes are only used for PSRP messages for a single remoting session.  PowerShell only closes the pipe at the end of the session at which time it also terminates the SSH process.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (PaulHigin) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: I am not sure what you mean, I only see this being used here in my branch.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (PaulHigin) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Sure, but I do see the point of DotNet CoreFx developers who originated this code.  The assert condition is pretty explanatory...

File: src/libpsl-native/src/createprocess.cpp
Comment (daxian-dbw) [inline src/libpsl-native/src/createprocess.cpp]: Without `close-on-exec`, the pipe file descriptors will be leaked to all subsequent processes created by powershell. The symptom I described above is one result this would cause. Although it's not applicable to the SSH remoting scenario, my worry is that there may be other symptoms resulted by the leak that would cause a problem we don't know yet which might be hard to diagnose.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: I think @iSazonov means that `s_utf8NoBom` should be put at some common place so that a single instance can be used in all places in our code base.

**Comment by daxian-dbw** [comment]
@iSazonov I see this in the man page of `SIGPROCMASK`
> The use of sigprocmask() is unspecified in a multithreaded process;

This makes me feel safer to use `sigactions`.

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs]: I was hoping to simply the code a bit between platforms.  But I see your point and will fix.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (PaulHigin) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Oh I see.  Yes this does seem like something that would be useful.  I'll move it to PsUtils.cs

File: src/libpsl-native/src/createprocess.cpp
Comment (PaulHigin) [inline src/libpsl-native/src/createprocess.cpp]: Ok, the leak definitely sounds bad.  I'll add the close-on-exec.  So that I understand more clearly, can you provide an example where the Fd is leaked even when SSH process is terminated?

Comment (daxian-dbw) [inline src/libpsl-native/src/createprocess.cpp]: Take the write end of the redirected stdin pipe as an instance (similar to the redirected stdout/stderr pipes), a FD pointing to the write end of that pipe will be created in powershell process after creating the SSH process. Assuming powershell then creates some other processes before SSH process terminates (I'm thinking of the scenario where I create a SSH PSSession but no an interactive remote session, so it's possible to run native commands locally while the SSH process is still alive), then that FD will be inherited by all those processes, so they are all potentially able to write to that pipe or affect that pipe in possible ways.

In this case, the SSH process is the only reader of that pipe, so when SSH terminates, "_the only reader dies, and then the writer gets notified about this by getting a SIGPIPE or at least an EPIPE error (depending on how signals are defined)._" (quoted from [here](https://stackoverflow.com/questions/19265191/why-should-you-close-a-pipe-in-linux))

I'm not saying it will definitely cause a problem, but just that if the leaked FD cause trouble to us, it might be very hard to figure out the real cause.

Comment (PaulHigin) [inline src/libpsl-native/src/createprocess.cpp]: Thanks, this definitely helps and could be an issue for the fan-out scenario.  I'll have this change in the PR update.

File: src/libpsl-native/src/createprocess.cpp
Comment (daxian-dbw) [inline src/libpsl-native/src/createprocess.cpp]: There are some additional changes for the close-on-exec -- calls to `CloseIfOpen` are removed from the `if (processId == 0)` block, see the CoreFx code [here](https://github.com/dotnet/corefx/blob/release/2.0.0/src/Native/Unix/System.Native/pal_process.cpp#L167).

I may be easier to see the differences in the PR change: https://github.com/dotnet/corefx/commit/6d204a6a83e3eca283535ca930157350d4adabd6

**Comment by daxian-dbw** [comment]
@PaulHigin Could you please update the PR description? A great discussion happened in this PR that led to a better solution. So the original description is out-dated and we'd better capture the solution of using `SIGPROCMASK` in the PR description.

**Comment by iSazonov** [comment]
LGTM.

Sorry for the long discussion - my last code for Unix was exactly 20 years ago. I'm glad I can remember just something. 😄

**Review by mirichmo — CHANGES_REQUESTED** [review]

File: src/libpsl-native/src/createprocess.cpp
Comment (PaulHigin) [inline src/libpsl-native/src/createprocess.cpp]: Fixed

**Comment by PaulHigin** [comment]
@iSazonov Thank you for your help on this.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (PaulHigin) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Done

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (PaulHigin) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Done

**Comment by SteveL-MSFT** [comment]
@iSazonov can you formally approve if you're good with the changes?  Thanks!

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@SteveL-MSFT Approved.

**Comment by PaulHigin** [comment]
@mirichmo Is there any reason why this cannot be merged?

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs
Comment (mirichmo) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: Please clean this up

**Review by mirichmo — CHANGES_REQUESTED** [review]

**Review by mirichmo — APPROVED** [review]

**Comment by PaulHigin** [comment]
@mirichmo  Thanks!

**Pull request merged**
